Fixed compilation with giflib version 5

Mailaender opened the pull request:
Based on https://github.com/pld-linux/libgdiplus/blob/master/libgdiplus-giflib.patch

monojenkins commented:
Hello! I'm the build bot for the Mono project. I need approval from a Mono team member to build this pull request. A team member should reply with "approve" to approve a build of this pull request, "whitelist" to whitelist this and all future pull requests from this contributor, or "build" to explicitly request a build, even if one has already been done. Contributors can ignore this message.

Mailaender commented:
Submitted at https://build.opensuse.org/request/show/295390

akoeplinger commented on src/gifcodec.c:
```diff
@@ -583,8 +579,11 @@ gdip_load_gif_image (void *stream, GpImage **image, BOOL from_file)
 	DGifCloseFile (gif);
-
+#else
+	DGifCloseFile (gif);
```
Both the if and the else do the same thing?

akoeplinger commented on src/gifcodec.c:
```diff
@@ -599,7 +598,11 @@ gdip_load_gif_image (void *stream, GpImage **image, BOOL from_file)
+#if (GIFLIB_MAJOR > 5) || ((GIFLIB_MAJOR == 5) && (GIFLIB_MINOR >= 1))
 		DGifCloseFile (gif);
+#else
+		DGifCloseFile (gif);
```
same here. I think you're missing the `, NULL` as in the original patch.

Mailaender commented:
Updated. https://build.opensuse.org/request/show/295392

akoeplinger commented:
Thanks. I don't have the authority to decide whether this is OK to be merged or not, can you reach out on the mailing list?

Mailaender commented:
Announced the fix at https://bugzilla.xamarin.com/show_bug.cgi?id=5859

akoeplinger commented on src/gifcodec.c:
```diff
@@ -306,9 +302,9 @@ gdip_load_gif_image (void *stream, GpImage **image, BOOL from_file)
+		gif = DGifOpen(stream, &gdip_gif_fileinputfunc, NULL);
 	} else {
-		gif = DGifOpen (stream, &gdip_gif_inputfunc);
+		gif = DGifOpen (stream, &gdip_gif_inputfunc, NULL);
```
these need the GIFLIB_MAJOR guards as well, or it will fail to compile on earlier giflibs.

akoeplinger commented on src/gifcodec.c:
```diff
@@ -663,9 +666,9 @@ gdip_save_gif_image (void *stream, GpImage *image, BOOL from_file)
+		fp = EGifOpenFileName (stream, 0, NULL);
 	} else {
-		fp = EGifOpen (stream, gdip_gif_outputfunc);
+		fp = EGifOpen (stream, gdip_gif_outputfunc, NULL);
```
same here

akoeplinger commented:
I just tried this out (with the added guards), but the "make check" tests fail with this patch

akoeplinger commented:
I added the additional guards here: https://github.com/akoeplinger/libgdiplus/commit/85fb4b4982b8c1f6d93e15982fea64e4e25efa11, feel free to pull that commit in. The libgdiplus tests run fine with it, but it'd be good to run the System.Drawing testsuite against it.

Mailaender commented:
Updated. Also added unit checks to https://build.opensuse.org/request/show/295418 however there is currently no current SUSE version that still runs giflib 4 so this remains untested from my side.

akoeplinger commented:
Thanks, looks good :+1: I tested on Ubuntu 14.04, which is still on giflib 4 and seems fine.

directhex commented:
build

directhex commented:
https://jenkins.mono-project.com/job/test-libgdiplus-pull-request/7/

Mailaender commented:
You should probably close https://bugzilla.xamarin.com/show_bug.cgi?id=5859 then.
